### PR TITLE
Synchronously send error response on message decoding failure

### PR DESCRIPTION
**Pull request by flowtoolz**
* Replaces two logs since those don't result in response messages anymore
* Adds a func for sending messages synchronously, to get the two messages out before fatalError

### CODE REVIEW

**Comment by flowtoolz** [comment]
@swift-ci please test

**Comment by MaxDesiatov** [comment]
@swift-ci please test

File: Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift [hunk 1]
```diff
@@ -303,7 +313,17 @@ public final class JSONRPCConnection {
...
   }
 
-  func send(encoding: (JSONEncoder) throws -> Data) {
+  private func sendMessageSynchronously(_ messageData: Data) {
+    let header = "Content-Length: \(messageData.count)\r\n\r\n"
+    guard let headerData = header.data(using: .utf8) else {
+      log("error encoding header", level: .error)
+      return
+    }
+    let stdOutHandle = FileHandle(fileDescriptor: sendIO.fileDescriptor)
```
Comment (benlangmuir) [inline Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift:322]: This seems unsafe since there may be other asynchronous writers.  How about something like this?

```swift
let semaphore = DispatchSemaphore(value: 0)
sendIO.barrier {
  ... write to the file descriptor, which we now have exclusive access to ...
  semaphore.signal()
}

_ = semaphore.wait(timeout: .now() + .seconds(3))
```

Although perhaps the specific timeout should be passed in.

File: Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift [hunk 2]
```diff
@@ -303,7 +313,17 @@ public final class JSONRPCConnection {
     send(_rawData: dispatchData)
   }
 
-  func send(encoding: (JSONEncoder) throws -> Data) {
+  private func sendMessageSynchronously(_ messageData: Data) {
+    let header = "Content-Length: \(messageData.count)\r\n\r\n"
+    guard let headerData = header.data(using: .utf8) else {
+      log("error encoding header", level: .error)
```
Comment (benlangmuir) [inline Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift:319]: This seems like it ought to be a fatalError.  There is no reason this should ever happen.

File: Sources/LanguageServerProtocolJSONRPC/MessageCoding.swift [hunk 3]
```diff
@@ -17,7 +17,7 @@ public enum JSONRPCMessage {
   case notification(NotificationType)
   case request(_RequestType, id: RequestID)
   case response(ResponseType, id: RequestID)
-  case errorResponse(ResponseError, id: RequestID)
```
Comment (benlangmuir) [inline Sources/LanguageServerProtocolJSONRPC/MessageCoding.swift:20]: We decode this value below like so:

```swift
      case (let id?, nil, _, let error?):
        msgKind = .response
        self = .errorResponse(error, id: id)
```

... which assumes that the `id` is non-nil.  That probably needs to get updated to `let id` instead of `let id?`.  Please add a test to `Tests/LanguageServerProtocolJSONRPCTests/CodingTests.swift` ensuring that we can round-trip this correctly when `id == nil`.

File: Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift [hunk 4]
```diff
@@ -172,6 +172,12 @@ public final class JSONRPCConnection {
     }
     return ready
   }
+  
+  private func send(_ message: JSONRPCMessage, async: Bool = true) {
```
Comment (benlangmuir) [inline Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift:176]: We should test the new `async == false` behaviour. We can underscore and make this public:

```swift
/// *Public for testing*
public func _send(...)
```

and add a test in `Tests/LanguageServerProtocolJSONRPCTests/ConnectionTests.swift` that uses this.  There are other tests that do similar things like sending raw data.

File: Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift [hunk 5]
```diff
@@ -229,17 +233,19 @@ public final class JSONRPCConnection {
               continue MESSAGE_LOOP
             }
           case .unknown:
-            log("error decoding message: \(error.message)", level: .error)
+            send(.errorResponse(ResponseError(error), id: nil),
```
Comment (benlangmuir) [inline Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift:236]: We can't test this behaviour directly due to the fatalError, but it might be good to add a test for the JSONRPCConnection receiving an error with `nil` id.  For example, if you manually send an `errorResponse(..., id: nil)` make sure it doesn't crash the receiving connection.  This would fit in `Tests/LanguageServerProtocolJSONRPCTests/ConnectionTests.swift`

**Review by benlangmuir — CHANGES_REQUESTED** [review]

File: Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift (hunk 5 above)
Comment (flowtoolz) [reply Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift]: With this one I better ask you before going down the wrong rabbit hole ... I don't see an easy way for testing that a `JSONRPCMessage`s or a concrete error response reaches the receiver. I'd want to write somethin like this:

```
func testSendErrorResponseWithNilID() {
    let client = connection.client

    let expectation = self.expectation(description: "received message")
    client.handleNextMessage { (message: JSONRPCMessage) in
      expectation.fulfill()
    }
  
    connection.serverConnection._send(.errorResponse(.cancelled, id: nil))

    waitForExpectations(timeout: 10)
  }
```

File: Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift (hunk 1 above)
Comment (flowtoolz) [reply Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift]: Tried to address this in https://github.com/apple/sourcekit-lsp/pull/334/commits/3f4c1c5c29a509809ab18d405b2ea0a7176afea7 without the `barrier` and preferring existing means ... I hope the `completionHandler` is ok?

File: Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift (hunk 2 above)
Comment (flowtoolz) [reply Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift]: done

File: Sources/LanguageServerProtocolJSONRPC/MessageCoding.swift (hunk 3 above)
Comment (flowtoolz) [reply Sources/LanguageServerProtocolJSONRPC/MessageCoding.swift:20]: done

File: Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift (hunk 4 above)
Comment (flowtoolz) [reply Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift]: done

File: Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift (hunk 5 above)
Comment (benlangmuir) [reply Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift]: Ah I see we don't have a good way to test this right now.  Let's skip this.

**Comment by benlangmuir** [comment]
@swift-ci please test

**Review by benlangmuir — COMMENTED** [review]
Looks good to me.  Please regenerate the tests for Linux (`swift test --generate-linuxmain`), and I think it will be ready to go :-)

**Comment by flowtoolz** [comment]
Done. Thanks for the learning opportunity 😊

**Comment by benlangmuir** [comment]
@swift-ci please test

**Comment by benlangmuir** [comment]
```
16:01:54 Test Case '-[LanguageServerProtocolJSONRPCTests.CodingTests testMessageDecodingError]' started.
16:01:54 
/Users/buildnode/jenkins/workspace/swift-sourcekit-lsp-PR-macOS/branch-main/sourcekit-lsp/Tests/LanguageServerProtocolJSONRPCTests/CodingTests.swift:161: error: -[LanguageServerProtocolJSONRPCTests.CodingTests testMessageDecodingError] : failed - expected error not seen
16:01:54 Test Case '-[LanguageServerProtocolJSONRPCTests.CodingTests testMessageDecodingError]' failed (1.289 seconds).
```

**Comment by flowtoolz** [comment]
Oh, I guess this is now a decodable error response message:

```
{"jsonrpc":"2.0","error":{"code":-32000,"message":""}}
```

So decoding that produces no error in CodingTests.swift:161. Should I remove the test or adjust it? I'm not sure what kind of LSP violation the test is testing for ...

**Comment by benlangmuir** [comment]
Sounds good to me

**Comment by benlangmuir** [comment]
@swift-ci please test

**Comment by flowtoolz** [comment]
done. this time, regenerating the Linux tests had no effect, just so you know 🤷🏼‍♂️
